gobblin-utility/src/main/java/gobblin/util/logs/LogCopier.java

```diff
@@ -0,0 +1,446 @@
+
+  // A set of current source log files being copied
+  private final Set<Path> srcLogFiles = Sets.newHashSet();
+
+  // A map from source log file paths to the most recent positions the copier have copied to
+  private final Map<Path, Long> currentPosMap = Maps.newHashMap();
+
+  private LogCopier(Builder builder) {
+    this.srcFs = builder.srcFs;
+    this.destFs = builder.destFs;
+
+    this.srcLogDir = this.srcFs.makeQualified(builder.srcLogDir);
+    this.destLogDir = this.destFs.makeQualified(builder.destLogDir);
+
+    this.sourceLogFileMonitorInterval = builder.sourceLogFileMonitorInterval;
+    this.copyInterval = builder.copyInterval;
+    this.timeUnit = builder.timeUnit;
+
+    this.logFileExtensions = builder.logFileExtensions;
+
+    this.filterRegexPattern = Optional.fromNullable(builder.filterRegexPattern);
+
+    this.logFileNamePrefix = Optional.fromNullable(builder.logFileNamePrefix);
+
+    this.linesWrittenBeforeFlush = builder.linesWrittenBeforeFlush;
+
+    this.logCopyExecutor = Executors.newScheduledThreadPool(0, ExecutorsUtils
+        .newThreadFactory(Optional.of(LOGGER), Optional.of("LogCopyExecutor")));
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    ExecutorsUtils.shutdownExecutorService(this.logCopyExecutor, Optional.of(LOGGER));
+  }
+
+  @Override
+  protected void runOneIteration() throws IOException {
+    checkSrcLogFiles();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(0, this.sourceLogFileMonitorInterval, this.timeUnit);
+  }
+
+  /**
+   * Perform a check on new source log files and submit copy tasks for new log files.
+   */
+  private void checkSrcLogFiles() throws IOException {
+    List<FileStatus> srcLogFiles = FileListUtils.listFilesRecursively(this.srcFs, this.srcLogDir, new PathFilter() {
+      @Override
+      public boolean accept(Path path) {
+        return logFileExtensions.contains(Files.getFileExtension(path.getName()));
+      }
+    });
+
+    if (srcLogFiles.isEmpty()) {
+      LOGGER.warn("No log file found under directory " + this.srcLogDir);
+      return;
+    }
+
+    Set<Path> currentSrcLogFiles = Sets.newHashSet();
+    for (FileStatus srcLogFile : srcLogFiles) {
+      currentSrcLogFiles.add(srcLogFile.getPath());
+    }
+
+    Set<Path> newLogFiles = Sets.newHashSet(currentSrcLogFiles);
+    synchronized (this.srcLogFiles) {
+      // Compute the set of new log files since the last check
+      newLogFiles.removeAll(this.srcLogFiles);
+      this.srcLogFiles.addAll(newLogFiles);
+    }
+
+    for (final Path srcLogFile : newLogFiles) {
+      String destLogFileName = this.logFileNamePrefix.isPresent() ?
+          this.logFileNamePrefix.get() + "." + srcLogFile.getName() : srcLogFile.getName();
+      final Path destLogFile = new Path(this.destLogDir, destLogFileName);
+
+      // Submit one copy task per new source log file
+      this.logCopyExecutor.scheduleAtFixedRate(new Runnable() {
```

Comment:
if `srcLogFile` is deleted, or the job that generates `srcLogFile` has finished, is there a way to remove the Runnable from schedule, or is it scheduled forever?